gammapy jupyter CLI to perform global actions on notebooks

Bultako opened the pull request:
This PR provides a new CLI ` gammapy jupyter` to perform global actions on a list of notebooks present in the current working dir, in a specific folder or for a specific file. The list of actions that provides are: strip output cells, execute, unittest to find broken notebooks and black-format.

<pre>
$ gammapy jupyter 
Usage: gammapy jupyter [OPTIONS] COMMAND [ARGS]...

  Perform a series of actions on Jupyter notebooks.

  The chosen action is applied for every Jupyter notebook present in the
  current working directory. Option --file allows to chose a single file,
  while option --fold allows to choose a different folder to scan. These
  options are mutually exclusive, only one is allowed.

  Examples
  --------
  
  $ gammapy jupyter stripout
  $ gammapy jupyter --src=mynotebooks.ipynb execute
  $ gammapy jupyter --src=myfolder/tutorials test
  $ gammapy jupyter black

Options:
  --src TEXT  Local folder or Jupyter notebook filename.
  -h, --help  Show this message and exit.

Commands:
  black     Format code cells with black.
  execute   Execute Jupyter notebooks.
  stripout  Strip output cells.
  test      Check if Jupyter notebooks are broken.
</pre>

cdeil commented on gammapy/scripts/jupyter.py:
```diff
@@ -0,0 +1,149 @@
+    return output
+
+
+def build_nblist(ctx):
```
Suggest to make this a generator function that `yield`s the paths. That means you can remove the lines that make the empty list and return it at the end, and the `nblist.append` calls become `yield` statements.
You can then use it as an iterable in the for loop directly, or call `list()` on the generator if you need a list, but i nthis case you don't need a list.

Bultako replied:
Done.

cdeil commented on gammapy/scripts/jupyter.py:
```diff
@@ -0,0 +1,149 @@
+    """Comment magic commands when formatting cells."""
+
+    lines = input.splitlines(True)
+    output = ""
```
Again, this would be little nicer as a generator function that yields the lines.

Bultako replied:
Well, we need the whole code in cell to send it to black, not only the lines.
I do not see how to a generator may help..

cdeil replied:
The generator helps a little bit, it reduces the number of lines you have here, and it's also a bit more efficient. Note that Python strings are immutable, so doing "+" repeatedly to build larger and larger strings isn't efficient.
If you yield strings for each line here, then in the caller you can do this:
```
txt = '\n'.join(comment_magics(input))
```
I think str.join will accept a generator that yields strings.

It's not a big improvement, just a very minor one.

Bultako replied:
Ok, I got it 👍 
Thanks for the review

cdeil commented on gammapy/scripts/jupyter.py:
```diff
@@ -0,0 +1,149 @@
+                ignorelist.append(nbname)
+        folder = notebook.parent
+
+        print('oups')
```
What's up with this print? Remove or make more clear?

Bultako replied:
Oups ! removed.

cdeil commented on gammapy/scripts/jupyter.py:
```diff
@@ -0,0 +1,149 @@
+        nbformat.write(nb, str(notebook))
+
+        # inform
+        print('Jupyter notebook {} painted in black.'.format(str(notebook)))
```
Does the mix of print statements and log messages have a logic to it?
Maybe change to `log` messages everywhere?

Bultako replied:
done.

cdeil commented on environment-dev.yml:
```diff
@@ -37,10 +37,10 @@ dependencies:
   - pandoc==2.1.3
   - pylint
   - flake8
+  - nbstripout
```
I'm surprised that it's useful to have a dependency to strip output cells.
Shouldn't it be < 5 lines to do this via the `nbconvert` or `nbformat` API to loop over output cells and make them empty?
Of course, if it's nontrivial / useful to use `nbstripout`, OK to add the dependency.

Bultako replied:
ok, done.
anyways, I have foreseen for developers to use `nbstripout` to configure a git filter before committing notebooks, so thought would be installed in the devel environment. But I agree, better to reduce dependencies.

cdeil commented on gammapy/scripts/jupyter.py:
```diff
@@ -55,42 +53,47 @@ def cli_jupyter_black(ctx):
+        # strip output cells
+        for cellnumber, cell in enumerate(nb.cells):
+            if nb.cells[cellnumber]['cell_type'] == 'code':
+                nb.cells[cellnumber]['outputs'] = []
```
Is this equivalent?
```
for cell in nb.cells:
    if cell['cell_type'] == 'code':
        cell['outputs'] = []
```

cdeil commented on gammapy/scripts/jupyter.py:
```diff
@@ -55,42 +53,47 @@ def cli_jupyter_black(ctx):
         try:
             t = time.time()
             subprocess.call(
                 "jupyter nbconvert --allow-errors --ExecutePreprocessor.timeout=None --ExecutePreprocessor.kernel_name=python3 --to notebook --execute '{}'.format(notebook) --inplace",
```
Can you please break this long string?
Suggest something like this:
```
subprocess.call(
    "jupyter nbconvert "
    "--allow-errors "
    "--ExecutePreprocessor.timeout=None "
...
```
Python will just concatenate the strings, and this makes it easy to read and to comment some option out or add code comments in between if something needs an explanation.

cdeil commented on gammapy/scripts/jupyter.py:
```diff
@@ -0,0 +1,144 @@
+                    semicolon = 0
+                    fmt = '\n'.join(comment_magics(fmt))
+                    #fmt = comment_magics(fmt)
+                    if fmt.endswith(';'):
```
I think you can just put:
```
has_semicolon = fmt.endswith(';')
```
and remove the if here and line `semicolon = 0` above.

cdeil commented on gammapy/scripts/jupyter.py:
```diff
@@ -0,0 +1,144 @@
+        # write formatted notebook
+        nbformat.write(nb, str(notebook))
+
+        # inform
```
remove comment

cdeil commented on gammapy/scripts/jupyter.py:
```diff
@@ -0,0 +1,144 @@
+                fmt = fmt.replace("###-MAGIC COMMAND-", "")
+            nb.cells[cellnumber]['source'] = fmt
+
+        # write formatted notebook
```
remove comment, just restates the next code line.

cdeil commented on gammapy/scripts/jupyter.py:
```diff
@@ -0,0 +1,144 @@
+                try:
+                    semicolon = 0
+                    fmt = '\n'.join(comment_magics(fmt))
+                    #fmt = comment_magics(fmt)
```
remove comment.

cdeil commented on gammapy/scripts/jupyter.py:
```diff
@@ -0,0 +1,144 @@
+def cli_jupyter_black(ctx):
+    """Format code cells with black."""
+
+    for notebook in build_nblist(ctx):
```
What you call `notebook` here are Path objects, not Notebook objects?
Maybe rename the variable to "path"?

Bultako commented:
@cdeil 

I have refactored most of the code and removed the `testipynb` dependency. The functionality for testing notebooks is coded in the CLI script `jupyter.py`, so no need to call functions in `gammapy.utils`. We could discuss later if we move all these functionalities to `gammapy.utils`, but I think that for the moment they could live in `gammapy jupyter` CLI, remove scripts `black_notebooks.py` and `test_notebooks.py`, and close PR https://github.com/gammapy/gammapy/pull/1765

Surely there are still fixes to address in code review, and I thank you for your patience.

cdeil commented on gammapy/scripts/jupyter.py:
```diff
@@ -0,0 +1,157 @@
+        log.error(ex)
+
+
+def read_notebook(path):
```
Is this `read_notebook` function useful?

You call it three times. In one case you do handle the "none" error case, in the other two you don't and will fail with a cryptic AttributeError in that case I think.

So this exception is for corrupt Jupyter notebook files? I think it would be OK to just let `nbformat.read` error out on those and halt, because in practice that doesn't occur, and if it does, it's clear that there's a corrupt file, no?

If you agree, you could just have the `nbformat.read` call in every current caller of `read_notebook` and delete this helper function.

cdeil commented on gammapy/scripts/jupyter.py:
```diff
@@ -0,0 +1,157 @@
+    lines = cellcode.splitlines(False)
+    for line in lines:
+        if line.startswith("%") or line.startswith("!"):
+            magic_line = "###-MAGIC COMMAND-" + line
```
This `tag_magics` function appears very far from where it's called, making the code overall not so easy to read. Also the magic string "###-MAGIC COMMAND-" shouldn't be duplicated in two places.

If blackening a notebook needs more than one function, i.e. a second helper function makes sense, I would suggest to group the two functions in a class. The task of the class would be to process one notebook., the loop over notebooks would remain in the click cli function.

cdeil commented on gammapy/scripts/jupyter.py:
```diff
@@ -0,0 +1,157 @@
+        run_notebook(path)
+        rawnb = read_notebook(path)
+
+        if rawnb:
```
I would always move the code black that does the processing into a function. In this case "test_notebook".
Then the CLI is just short, doing the loop over notebooks, read, process, write.

cdeil commented on gammapy/scripts/main.py:
```diff
@@ -83,6 +83,40 @@ def cli_download(ctx, folder):
+        'fold': fold,
+    }
+
+    if file and fold != '.':
```
Can't you just have one option (e.g. "src" like black, see https://github.com/ambv/black#command-line-options), and then do:
```
path = Path(src)
if path.isdir():
   paths = ...
else:
    paths = [path]
```
and the rest of the code always gets a list of paths to process?

I think that might be a bit simpler?

cdeil replied:
If you do this you could just move the code that's currently in build_nb in called ~ 3 times into the main cli and store `paths` on the ctx, and then just do `for path in ctx.paths` for each subcommand.

This is how I would do it, but it's not very different from what you have, just a matter of taste. Do as you prefer.

cdeil commented:
@Bultako - Thank you for your work here. I think this is turning out nicely, very simple to read and understand and maintain code that does what we need with notebooks. 

You have to delay the imports of e.g. "black" or "jupyter" to the functions, so that `gammapy info` works for users without those "dev dependencies" installed. See e.g. here:
https://ci.appveyor.com/project/cdeil/gammapy/build/1.0.4447/job/8iuh1p09wv7g8ed6#L1552

Once CI passes, feel free to merge anytime.

cdeil commented on gammapy/scripts/jupyter.py:
```diff
@@ -0,0 +1,166 @@
+log = logging.getLogger(__name__)
+
+try:
+    import nbformat
```
You can't have this at the top of the module, it needs to be a delayed import in the functions that use this.
See https://ci.appveyor.com/project/cdeil/gammapy/build/1.0.4455/job/qfhona74oc6yy5sw#L1569

I would just let the ImportError occur. We do this even for users e.g. in functions that use scipy or matplotlib in Gammapy, so even more so for dev tools I think it's OK to see an ImportError occur.

Bultako replied:
What do you mean by a 'delayed import'? Am I supposed to leave the imports even if CI fails? I was struggling to fix the CI by any means.

cdeil replied:
With delayed import I mean to put the import into the function where you use it on the first line. I don't know if there's a proper term for that.

Then CI and tests will not fail if they just import the module.
And when writing tests for code that uses an optional dependency, we have the `requires_dependency` decorator that we use to mark each such test function.

cdeil commented on gammapy/scripts/jupyter.py:
```diff
@@ -0,0 +1,166 @@
+@click.pass_context
+def cli_jupyter_black(ctx):
+    """Format code cells with black."""
+    try:
```
Same here, I'd suggest to remove the try / except and just `import black`.

cdeil commented on gammapy/scripts/jupyter.py:
```diff
@@ -0,0 +1,166 @@
+                fmt = fmt.replace(self.MAGIC_TAG, "")
+            cell['source'] = fmt
+
+    def tag_magics(self, cellcode):
```
Make this a `@staticmethod` since self isn't accessed?

Bultako replied:
There is `self.MAGIC_TAG`
How about moving `self.MAGIC_TAG` out of the function and have it s a global var in `jupyter.py`? At this moment we could have @staticmethod

cdeil replied:
No, then just leave as-is. It's perfectly fine, I just didn't see the access to `self` in this method.

cdeil replied:
Better to have everything on this one class. Then it's easier to maintain and others can copy & paste it.

Actually, I think a separate `blacknb` or `black-notebook` Python package that does this would find a userbase; or we could file an issue in one of the Jupyter repos asking if they think it's appropriate to be included there directly.

cdeil commented on gammapy/scripts/main.py:
```diff
@@ -83,26 +85,75 @@ def cli_download(ctx, folder):
+        sys.exit()
+
+    if path.is_dir():
+        for f in path.iterdir():
```
I think this would be equivalent and a tiny bit shorter:
```
if path.is_dir():
    path = path.glob('*.ipynb')
```

cdeil replied:
Probably making a list instead of a generator here is less error-prone (e.g. if it's looped over twice a generator will be empty on the second go):
```
if path.is_dir():
    paths = list(path.glob('*.ipynb'))
```
